### ISSUE TITLE
Snuba consumers keep crashing

### ISSUE THREAD

**Issue by zifeo** [sentry-kubernetes/charts#365]
sentry-snuba-consumer & sentry-snuba-outcomes-consumer restart from time to time and I unable to find was could cause the issue upstream. The logs show the following:

```
2021-04-27 08:09:49,628 Error submitting packet, dropping the packet and closing the socket
2021-04-27 08:09:49,632 Completed processing <Batch: 1 message, open for 1.04 seconds>.
2021-04-27 08:09:50,897 Completed processing <Batch: 1 message, open for 1.05 seconds>.
2021-04-27 08:09:57,980 Error submitting packet, dropping the packet and closing the socket
2021-04-27 08:09:57,985 Completed processing <Batch: 1 message, open for 1.04 seconds>.
2021-04-27 08:10:00,374 Error submitting packet, dropping the packet and closing the socket
2021-04-27 08:10:00,375 Error submitting packet, dropping the packet and closing the socket
2021-04-27 08:10:00,378 Completed processing <Batch: 5 messages, open for 1.28 seconds>.
2021-04-27 08:10:03,380 Caught ProtocolError('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer')), shutting down...
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/urllib3/connectionpool.py", line 600, in urlopen
    httplib_response = self._make_request(conn, method, url,
  File "/usr/local/lib/python3.8/site-packages/urllib3/connectionpool.py", line 387, in _make_request
    six.raise_from(e, None)
  File "<string>", line 2, in raise_from
  File "/usr/local/lib/python3.8/site-packages/urllib3/connectionpool.py", line 383, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/local/lib/python3.8/site-packages/sentry_sdk/integrations/stdlib.py", line 102, in getresponse
    rv = real_getresponse(self, *args, **kwargs)
  File "/usr/local/lib/python3.8/http/client.py", line 1344, in getresponse
    response.begin()
  File "/usr/local/lib/python3.8/http/client.py", line 307, in begin
    version, status, reason = self._read_status()
  File "/usr/local/lib/python3.8/http/client.py", line 268, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/usr/local/lib/python3.8/socket.py", line 669, in readinto
    return self._sock.recv_into(b)
ConnectionResetError: [Errno 104] Connection reset by peer

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/snuba", line 33, in <module>
    sys.exit(load_entry_point('snuba', 'console_scripts', 'snuba')())
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/src/snuba/snuba/cli/consumer.py", line 161, in consumer
    consumer.run()
  File "/usr/src/snuba/snuba/utils/streams/processing/processor.py", line 112, in run
    self._run_once()
  File "/usr/src/snuba/snuba/utils/streams/processing/processor.py", line 147, in _run_once
    self.__processing_strategy.poll()
  File "/usr/src/snuba/snuba/utils/streams/processing/strategies/streaming/transform.py", line 61, in poll
    self.__next_step.poll()
  File "/usr/src/snuba/snuba/utils/streams/processing/strategies/streaming/collect.py", line 122, in poll
    self.__close_and_reset_batch()
  File "/usr/src/snuba/snuba/utils/streams/processing/strategies/streaming/collect.py", line 104, in __close_and_reset_batch
    self.__batch.close()
  File "/usr/src/snuba/snuba/utils/streams/processing/strategies/streaming/collect.py", line 64, in close
    self.__step.close()
  File "/usr/src/snuba/snuba/consumers/consumer.py", line 225, in close
    self.__insert_batch_writer.close()
  File "/usr/src/snuba/snuba/consumers/consumer.py", line 90, in close
    self.__writer.write(
  File "/usr/src/snuba/snuba/clickhouse/http.py", line 245, in write
    batch.join()
  File "/usr/src/snuba/snuba/clickhouse/http.py", line 181, in join
    response = self.__result.result(timeout)
  File "/usr/local/lib/python3.8/concurrent/futures/_base.py", line 444, in result
    return self.__get_result()
  File "/usr/local/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
    raise self._exception
  File "/usr/local/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.8/site-packages/urllib3/connectionpool.py", line 640, in urlopen
    retries = retries.increment(method, url, error=e, _pool=self,
  File "/usr/local/lib/python3.8/site-packages/urllib3/util/retry.py", line 368, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/usr/local/lib/python3.8/site-packages/urllib3/packages/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "/usr/local/lib/python3.8/site-packages/urllib3/connectionpool.py", line 600, in urlopen
    httplib_response = self._make_request(conn, method, url,
  File "/usr/local/lib/python3.8/site-packages/urllib3/connectionpool.py", line 387, in _make_request
    six.raise_from(e, None)
  File "<string>", line 2, in raise_from
  File "/usr/local/lib/python3.8/site-packages/urllib3/connectionpool.py", line 383, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/local/lib/python3.8/site-packages/sentry_sdk/integrations/stdlib.py", line 102, in getresponse
    rv = real_getresponse(self, *args, **kwargs)
  File "/usr/local/lib/python3.8/http/client.py", line 1344, in getresponse
    response.begin()
  File "/usr/local/lib/python3.8/http/client.py", line 307, in begin
    version, status, reason = self._read_status()
  File "/usr/local/lib/python3.8/http/client.py", line 268, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/usr/local/lib/python3.8/socket.py", line 669, in readinto
    return self._sock.recv_into(b)
urllib3.exceptions.ProtocolError: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))
```

Are the regular error in the logs expected? I would expect to be a connectivity issue with clickhouse, but nothing is wrong on that end.

This is on GKE 1.19.8-gke.1600 with chart 11.2.0 (but has existed since a long time).

**Comment by oikonomou** [comment]
Is there any update on this? I've been facing the same error for long. This error appeared while using Sentry chart 9.0.2. Then upgraded to 10.1.0 but still there.

**Comment by RyanSiu1995** [comment]
We encountered the same situation across all the consumers. 
Eventually, by scaling up the consumers, the connection reset issue can be resolved.
If there is still restart for your consumers, that means you need more replicas in this case.
It is highly related to the implementation to the snuba consumer IMO.

**Comment by aamarques** [comment]
Any news from the Sentry Magic guys :)
 @BFK  @fpacifici

**Comment by fpacifici** [comment]
Hi, 
while sentry-kubernetes is not actually maintained or supported by Sentry, I think I can provide some details about the error.

```
File "/usr/src/snuba/snuba/clickhouse/http.py", line 181, in join
    response = self.__result.result(timeout)
```
This in the stack trace means that indeed the connection reset comes from Clickhouse, which seems like a connectivity issue with Clickhouse. 
I am not sure what could solve this by adding consumer replicas, but if you are running many Kafka partitions and a lot of load in term of events per second, adding replicas will reduce the size of the batches you write to Clickhouse and increase the frequency. (One batch per second is a good time, so it does not seem an issue for you).
Anyway it seems to me this kubernetes config sets the batch size here: https://github.com/sentry-kubernetes/charts/blob/develop/sentry/templates/deployment-snuba-consumer.yaml#L73-L77

```
2021-04-27 08:10:00,374 Error submitting packet, dropping the packet and closing the socket
```
These are unrelated and not critical. It simply means the consumer cannot connect itself to sentry to send its logging events to a sentry instance.
It may be expected depending whether you set a sentry DSN in the snuba config.

How often does this happen and does the consumer restarts and recover cleanly ?

**Comment by aamarques** [comment]
Thanks, @fpacifici and sorry but I didn't notice that this topic is about Kubernetes.
The error that occurs here is the same (related to clickhouse) but I'm using docker only.
Occurs one or two times a day with 5k/7k of total transactions. 
Also, I'm using the sentry default configuration. do you think that the proposed solution helps me?

**Comment by yayapao** [comment]
I've been facing the same error for long, and i found that CPU/memory consumption is usually not high. Seems that scaling up the consumers cannot help with it. Do you think that the proposed solution helps me?

**Comment by Mokto** [comment]
This issue is stale because it has been open for 30 days with no activity.

**Comment by zifeo** [comment]
@Mokto unstale please

**Comment by Mokto** [comment]
This issue is stale because it has been open for 30 days with no activity.

**Comment by arkdchst** [comment]
I managed to fix this by increasing `keep_alive_timeout` option in Clickhouse.
It seems like a bug in Snuba. When Snuba isn't using Clickhouse connection for certain amount of time, Clickhouse closes it. Why this causes Snuba restart I don't know.
With higher values of `keep_alive_timeout` Snuba containers continue restarting, but much more rarely.

**Comment by flixx** [comment]
Hello,

I was looking again into this issue. `keep_alive_timeout` indeed gave some improvements. But setting all snuba replicas to 2 was making the issue completely disappear for me.

I have this in my values now:

```
asHook: true

ingress:
  enabled: true
  ingressClassName: "traefik"
  regexPathStyle: "traefik"
nginx:
  enabled: false
sentry:
  worker:
    replicas: 1
  web:
    strategyType: Recreate
  subscriptionConsumerEvents:
    commitBatchSize: 100
rabbitmq:
  replicaCount: 1
redis:
  cluster:
    enabled: false
postgresql:
  persistence:
    size: 128Gi
kafka:
  persistence:
    size: 64Gi

snuba:
  api:
    replicas: 2
  consumer:
    replicas: 2
  outcomesConsumer:
    replicas: 2
  replacer:
    replicas: 2
  subscriptionConsumerEvents:
    replicas: 2
  subscriptionConsumerTransactions:
    replicas: 2
  sessionsConsumer:
    replicas: 2
  transactionsConsumer:
    replicas: 2


clickhouse:
  clickhouse:
    configmap:
      keep_alive_timeout: "20"
```

Using this settings I don't have restarts anymore.

@arkdchst You could also try to increase the snuba replicas.

==Update==

I still have restarts :(
Just less often. 

I guess it would be helpful if this chart could use a more up-to-date clickhouse version... The version that is used in the chart right now is from 2019...

https://github.com/sentry-kubernetes/charts/blob/develop/clickhouse/Chart.yaml#L2

Linked issue: https://github.com/sentry-kubernetes/charts/issues/660

**Comment by arkdchst** [comment]
@flixx 
Thank you. I have tried to increase number of pods up to 6 replicas. For my installation it didn't help.

My Clickhouse version is 21.8.10, it still works with current Sentry version, but problem still persists. To my mind, root of the problem is a bug in Snuba.

**Comment by Mokto** [comment]
This issue is stale because it has been open for 30 days with no activity.

**Comment by Mokto** [comment]
This issue was closed because it has been inactive for 14 days since being marked as stale.

**Comment by begemotik** [comment]
Encountered the issue as well

**Comment by Natgho** [comment]
Encountered the issue as well +1

**Comment by lekhanh1234** [comment]
TLDR: if you don't use Datadog, then it's fine

The error message: `Error submitting packet, dropping the packet and closing the socket` is originated from datadog library. 

This is where the error message comes from: [Datadog code](https://github.com/DataDog/datadogpy/blob/0873f87c96f72c4e4cbec56457e8d9a33e9f38b6/datadog/dogstatsd/base.py#L885-L886)

This is why sentry shows the error: [Sentry code ](https://github.com/getsentry/sentry/blob/f711263a91315352b5da763eb6901563a9187af8/src/sentry/metrics/dogstatsd.py#L3)

**Comment by kong62** [comment]
```
2023-06-06 18:35:33,735 Time limit reached, closing <Batch: 80 messages, open for 1.01 seconds>...
2023-06-06 18:35:33,736 Finished sending data from <HTTPWriteBatch: 80 rows (27875 bytes)>.
2023-06-06 18:35:33,759 http://mc-storage-ch.hupu.io:8123 "POST /?load_balancing=in_order&insert_distributed_sync=1&query=INSERT+INTO+sentry.sessions_raw_dist++FORMAT+JSONEachRow HTTP/1.1" 200 None
2023-06-06 18:35:33,760 Received response for <HTTPWriteBatch: 80 rows (27875 bytes)>.
2023-06-06 18:35:33,760 Waited 0.0248 seconds for 80 rows to be written to <HTTPBatchWriter: sentry.sessions_raw_dist on mc-storage-ch.hupu.io:8123>.
2023-06-06 18:35:33,760 Committing offsets: {Partition(topic=Topic(name='ingest-sessions'), index=8): Position(offset=2619519834, timestamp=datetime.datetime(2023, 6, 6, 18, 35, 33, 718000)), Partition(topic=Topic(name='ingest-sessions'), index=9): Position(offset=2619432141, timestamp=datetime.datetime(2023, 6, 6, 18, 35, 33, 691000))}
2023-06-06 18:35:33,761 Waited 0.0009 seconds for offsets to be committed to <arroyo.backends.kafka.consumer.KafkaConsumer object at 0x7fe82e7e22e0>.
2023-06-06 18:35:33,761 Completed processing <Batch: 80 messages, open for 1.03 seconds>.

2023-06-06 18:35:34,737 Time limit reached, closing <Batch: 85 messages, open for 1.00 seconds>...
2023-06-06 18:35:34,738 Finished sending data from <HTTPWriteBatch: 85 rows (29612 bytes)>.
2023-06-06 18:35:34,759 http://mc-storage-ch.hupu.io:8123 "POST /?load_balancing=in_order&insert_distributed_sync=1&query=INSERT+INTO+sentry.sessions_raw_dist++FORMAT+JSONEachRow HTTP/1.1" 200 None
2023-06-06 18:35:34,759 Received response for <HTTPWriteBatch: 85 rows (29612 bytes)>.
2023-06-06 18:35:34,759 Waited 0.0217 seconds for 85 rows to be written to <HTTPBatchWriter: sentry.sessions_raw_dist on mc-storage-ch.hupu.io:8123>.
2023-06-06 18:35:34,759 Committing offsets: {Partition(topic=Topic(name='ingest-sessions'), index=8): Position(offset=2619519877, timestamp=datetime.datetime(2023, 6, 6, 18, 35, 34, 714000)), Partition(topic=Topic(name='ingest-sessions'), index=9): Position(offset=2619432183, timestamp=datetime.datetime(2023, 6, 6, 18, 35, 34, 722000))}
2023-06-06 18:35:34,760 Waited 0.0010 seconds for offsets to be committed to <arroyo.backends.kafka.consumer.KafkaConsumer object at 0x7fe82e7e22e0>.
2023-06-06 18:35:34,760 Completed processing <Batch: 85 messages, open for 1.02 seconds>.

2023-06-06 18:35:35,753 Time limit reached, closing <Batch: 74 messages, open for 1.02 seconds>...
2023-06-06 18:35:35,754 Finished sending data from <HTTPWriteBatch: 74 rows (25763 bytes)>.

2023-06-06 18:35:36,764 Time limit reached, closing <Batch: 66 messages, open for 1.01 seconds>...

2023-06-06 18:35:41,764 Caught TimeoutError(), shutting down...
2023-06-06 18:35:41,765 Terminating <arroyo.processing.strategies.streaming.transform.TransformStep object at 0x7fe82e7e29a0>...
2023-06-06 18:35:41,765 Terminating <arroyo.processing.strategies.streaming.collect.ParallelCollectStep object at 0x7fe82e7e2640>...
2023-06-06 18:35:41,765 Terminating <snuba.consumers.consumer.ProcessedMessageBatchWriter object at 0x7fe82e72ec10>...
2023-06-06 18:35:41,765 Closing <arroyo.backends.kafka.consumer.KafkaConsumer object at 0x7fe82e7e22e0>...
2023-06-06 18:35:41,765 Partitions revoked: [Partition(topic=Topic(name='ingest-sessions'), index=8), Partition(topic=Topic(name='ingest-sessions'), index=9)]
2023-06-06 18:35:41,766 Processor terminated
Traceback (most recent call last):
  File "/usr/local/bin/snuba", line 33, in <module>
    sys.exit(load_entry_point('snuba', 'console_scripts', 'snuba')())
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/usr/src/snuba/snuba/cli/consumer.py", line 188, in consumer
    consumer.run()
  File "/usr/local/lib/python3.8/site-packages/arroyo/processing/processor.py", line 121, in run
    self._run_once()
  File "/usr/local/lib/python3.8/site-packages/arroyo/processing/processor.py", line 156, in _run_once
    self.__processing_strategy.poll()
  File "/usr/local/lib/python3.8/site-packages/arroyo/processing/strategies/streaming/transform.py", line 63, in poll
    self.__next_step.poll()
  File "/usr/local/lib/python3.8/site-packages/arroyo/processing/strategies/streaming/collect.py", line 135, in poll
    self.close_and_reset_batch()
  File "/usr/local/lib/python3.8/site-packages/arroyo/processing/strategies/streaming/collect.py", line 199, in close_and_reset_batch
    self.future.result(timeout=5)
  File "/usr/local/lib/python3.8/concurrent/futures/_base.py", line 446, in result
    raise TimeoutError()
concurrent.futures._base.TimeoutError
2023-06-06 18:35:42,121 http://mc-storage-ch.hupu.io:8123 "POST /?load_balancing=in_order&insert_distributed_sync=1&query=INSERT+INTO+sentry.sessions_raw_dist++FORMAT+JSONEachRow HTTP/1.1" 200 None
2023-06-06 18:35:42,121 Received response for <HTTPWriteBatch: 74 rows (25763 bytes)>.
2023-06-06 18:35:42,122 Waited 6.3678 seconds for 74 rows to be written to <HTTPBatchWriter: sentry.sessions_raw_dist on mc-storage-ch.hupu.io:8123>.
2023-06-06 18:35:42,122 Committing offsets: {Partition(topic=Topic(name='ingest-sessions'), index=8): Position(offset=2619519914, timestamp=datetime.datetime(2023, 6, 6, 18, 35, 35, 719000)), Partition(topic=Topic(name='ingest-sessions'), index=9): Position(offset=2619432220, timestamp=datetime.datetime(2023, 6, 6, 18, 35, 35, 734000))}
```

```
# cat /etc/clickhouse-server/config.xml |grep keep_alive_timeout
    <keep_alive_timeout>600</keep_alive_timeout>
```

any?

**Comment by makarov-roman** [comment]
Spend a day on that and figured out the problem I'm having. Not sure if it's poor API from Clickhouse side or poor error handling on Snuba, but if memory allocation fails on Clickhouse side - it will reset the connection (e.g. throws Network Error for non network situation).

I increased memory request/limits on clickhouse pod to 3.5Gi and now all consumers are running without resets.
